### ISSUE TITLE
Infinite "duplicate coordinates found" warning when using Haversine distance

### ISSUE THREAD

**Issue by Datseris** [JuliaEarth/Variography.jl#50]
MWE:

```julia
using GeoStats, Distances
lons = 0.5:10:359.5
lats = -89.5:10:89.5
C = rand(length(lons), length(lats))
D = copy(C)

data = georef((C = C, D = D); origin = (lons[1], lats[1]), spacing = (step(lons), step(lats))

@time γ = EmpiricalVariogram(data, :C; algo = :full, distance = Distances.Haversine())
```

It takes a while, but when the algorithm reaches the maximum latitude, it starts spiting out the warning.

I believe I have found out why this happens. Having a look at `data.domain` I see:

```
julia> data
648 MeshData{2,Float64}
  variables (rank 2)
    └─C (Float64)
    └─D (Float64)
  domain: 36×18 CartesianGrid{2,Float64}

julia> data.domain
ERROR: type NamedTuple has no field domain
Stacktrace:
 [1] getproperty
   @ .\Base.jl:33 [inlined]
 [2] getproperty
   @ ~\.julia\packages\TypedTables\WimSA\src\Table.jl:85 [inlined]
 [3] getcolumn
   @ ~\.julia\packages\Tables\YzCZp\src\Tables.jl:100 [inlined]
 [4] getindex(data::MeshData{CartesianGrid{2, Float64}, Dict{Int64, TypedTables.Table{NamedTuple{(:C, :D), Tuple{Float64, Float64}}, 1, NamedTuple{(:C, :D), Tuple{Vector{Float64}, Vector{Float64}}}}}}, col::Symbol)
   @ Meshes ~\.julia\packages\Meshes\JOSiV\src\traits\data.jl:144
 [5] getproperty(data::MeshData{CartesianGrid{2, Float64}, Dict{Int64, TypedTables.Table{NamedTuple{(:C, :D), Tuple{Float64, Float64}}, 1, NamedTuple{(:C, :D), Tuple{Vector{Float64}, Vector{Float64}}}}}}, col::Symbol)
   @ Meshes ~\.julia\packages\Meshes\JOSiV\src\traits\data.jl:151
 [6] top-level scope
   @ REPL[6]:1

julia> getfield(data, fieldnames(typeof(data))[1])
36×18 CartesianGrid{2,Float64}
  minimum: Point(0.5, -89.5)
  maximum: Point(360.5, 90.5)
  spacing: (10.0, 10.0)
```

(BTW not sure why `data.domain` doesn't work)

In any case, as you can see, the "maximum" point and the "minimum" point are in fact the exact same point in longitude-latitude as far as the Haversine distance is concerned. What I do not fully understand is why this point exists. Why isn't the maximum the actual maximum, e.g. `(lons[end], lats[end])`?

**Comment by juliohm** [comment]
> It takes a while, but when the algorithm reaches the maximum latitude, it starts spiting out the warning.

The ominidirectional variogram usually takes more time because it is averaging over all directions. One trick to get results more quickly is to use the `DirectionalVariogram` instead for lat and lon separately. In this case, we will use multiple threads to parallelize the calculation (lat lines are all parallel and can be processed in parallel, similar argument can be made for lon).

> (BTW not sure why data.domain doesn't work)

We use the `getproperty` to access variables in the geospatial data object because it is a Tables.jl table. So you can do `data.D` to get the result variable as a raw column vector, and even `data.geometry` to get the vector of quadrangles associated with this dataset. Given that the syntax is reserved for that purpose we ask users to use the public API to access information like the domain. You can do `domain(data)` to get the domain and `values(data)` to get the corresponding table.

> as you can see, the "maximum" point and the "minimum" point are in fact the exact same point in longitude-latitude as far as the Haversine distance is concerned. What I do not fully understand is why this point exists

That is the tricky part. When I mentioned on Slack that handling this use case properly would require a different topology I meant that a grid over a sphere is different than a Cartesian grid and the begin and end of the lat/lon arrays are treated differently in a sphere versus in a flat Cartesian world. The ultimate solution will be available in Meshes.jl soon when we introduce a SphereTopology. Right now we only have FullTopology for representing arbitrary meshes, and HalfEdgeTopology for representing orientable 2-manifold meshes. Your case could be handled in theory by both topologies, we would just need to replace the CartesianGrid by a spherical domain. If you would like to help in this direction, we could discuss over in Meshes.jl how to add this functionality. It would be very nice to have that. I am working on the GridTopology next week for some optimizations, but could take a look after 2 weeks in the spherical case.

**Comment by juliohm** [comment]
What I would do here given the current functionality available is try to project the lat/lon into a projection that preserves length and compute the variograms there. The variogram calculation itself has a lot of Euclidean assumptions baked in and we are far from generalizing it to general manifolds.

**Comment by Datseris** [comment]
Alright thanks. Yes I think I should go with your suggestion. Actually I Will just use a `PointSet` that has 3D points of Lon-Lat projected into Euclidean3D on the unit sphere.
